ScanDoc SDK implementation

avargic opened the pull request:
- Add ScanDoc support to monriSDK -  for scanning cards to extract data:

1.  ScanDocApi, ScanDocHttpApiImpl - entry point for ScanDoc related functions, performs network calls
2. ScanDocAuthenticationManager - for fetching and refreshing access tokens
3. ScanDocUrlProvider - provides backend urls
4. ScanDocExtractionConfig - configuration options for the "extraction" network call 
5. ScanDocApiOptions - for configuring ScanDoc backend 
6. ScanDoc request/response models 

- Move network requests from MonriHttpApiImpl to new HttpsClient class

avargic committed:
working on scandoc api

avargic committed:
working on scandoc api

avargic committed:
Merge branch 'development' into feature/scandoc

avargic committed:
Merge remote-tracking branch 'origin/feature/scandoc' into feature/scandoc

avargic committed:
Finishing scandoc implementation
Add scandoc functionality to test app
AGP & gradle version bump
Add cameraX dependency

avargic committed:
Add unit tests
Add kotlin configuration
Add assertJ dependency
Minor fixes
Add missing getters
Add missing response json properties

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/HttpsClient.java:
```diff
@@ -0,0 +1,185 @@
+
+    protected MonriHttpResult<JSONObject> httpsPOST(
+            final String endpoint,
+            final JSONObject body,
```
Možda bih apstrahirao body da ne mora nužno biti `JSONObject` već običan `String`.
Ovako se vežeš za org.json library i to da će svaki POST request imati body json formata, što uglavnom je slučaj, ali ne mora biti uvijek.

Ako si želiš pojednostavniti formatiranje/type request bodya i response bodya, možeš napraviti nekakav generic wrapper oko clienta koji će requeste nekog generic tipa mapirati kao String i response natrag na taj tip, pa bi u ovom slučaju to bilo JSONObject, a u slučaju ScanDoc-a bio onaj custom tip. Isto tako možeš dodati neki zajednički error tip.

Ako ne vidiš potrebu za tim, svakako bi base implementaiju HttpsClienta prebacio na čisti String, a onda proširiti sa JSONObjectom izvan te klase.

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/HttpsClient.java:
```diff
@@ -0,0 +1,185 @@
+
+            return readResponseFromInputStream(urlConnection);
+
+        } catch (Exception e) {
```
`final` hehe

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/HttpsClient.java:
```diff
@@ -0,0 +1,185 @@
+            final Map<String, String> headers
+    ) throws IOException {
+        URL url = new URL(endpoint);
+        HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
```
`final`

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/HttpsClient.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    protected void addHeadersToConnection(final HttpURLConnection urlConnection, final Map<String, String> headers) {
+        for (String key : headers.keySet()) {
```
final

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/HttpsClient.java:
```diff
@@ -0,0 +1,185 @@
+    private static void writeToOutputStream(final HttpURLConnection urlConnection, final JSONObject body) throws IOException {
+        OutputStreamWriter wr = null;
+
+        try {
```
try with resources ?

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriHttpApiImpl.java:
```diff
@@ -72,9 +64,9 @@ public MonriHttpResult<ConfirmPaymentResponse> confirmPayment(@NonNull ConfirmPa
         try {
-            final MonriHttpResult<JSONObject> response = httpsPOST(
+            final MonriHttpResult<JSONObject> response = httpsClient.httpsPOST(
                     baseUrl + "/v2/google-pay/" + paymentId + "/start-payment",
```
Sori, znam da kasnim jer je to od prošle verzije, ali ove url-ove ako možeš prebaciti u neku klasu kao što ima ScanDoc url provider 😬

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/ScanDocApiImpl.java:
```diff
@@ -0,0 +1,64 @@
+                TERMS_AND_CONDITIONS_ACCEPTED,
+                image.toString(),
+                skipImageSizeCheck,
+                List.of());
```
Ako ova lista treba biti i ostati prazna, možda bolje `Collections.emptyList()`, tako je jasna namjera da treba biti prazno (i mrvicu je brže tako 😄 )

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/ScanDocApiImpl.java:
```diff
@@ -0,0 +1,64 @@
+                    final MonriHttpResult<ScanDocValidateResponse> result = scanDocHttpApiImpl.validateScannedCard(request);
+
+                    if (result.getCause() != null) {
+                        throw result.getCause();
```
Nisam ljubitelj ovakvog handleanja grešaka. Točnije hvatanje exceptiona, pa bacanja, pa opet hvatanja exceptiona. 
Obzirom da je SDK tako posložen već, nemaš baš izbora pa nek stoji, samo želim napomenuti da to nije idealno i treba izbjegavati.

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/ScanDocAuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,73 @@
+    private final HttpsClient httpsClient;
+    private final String userKey;
+    private final String subClient;
+    private AccessToken accessToken;
+    private AccessToken refreshToken;
```
Izbjegavati držanje osjetljivih podataka u memoriji. `AccessToken` nije kritičan pa može biti u memoriji.
Nisam previše upućen u kontekst ScanDoc-a, ali rekao bih da access token ovdje nije rizična komponenta pa ne moraš mijenjati, ali komentiram ti da čisto da imaš na umu za idući put.

Ako se koriste neki osjetljiviji podaci, kao lozinke, api ključevi itd. onda ih je uvijek bolje spremati kriptirane (bilo u radnoj memoriji, keystoreu ili memoriji uređaja) i onda neposredno prije upotrebe dekriptirati.
Ako neki podatak koji služi za autentikaciju/autorizaciju treba biti u memoriji na dulji vremenski period, a nije isplativo ili potrebno raditi kriptiranje/dekriptiranje, bolje da nije `String` tipa jer je tako lako primjetan i čitljiv prilikom memory dumpova ili ostalih dinamičkih napada. Bolje ga je spremati kao byte ili char array.

luka-dragutin-asee reviewed (commented):
Nisam svugdje pisao, na nekoliko mjesta fali `final` pa prođi jednom i nadopuni haha

avargic committed:
Rename .java to .kt

avargic committed:
Add missing final modifiers
Refactoring httpsClient to return HttpsRequest
Add httpsClientProxy
Moving Monri urls to url provider
Configure kotlin for application module

avargic committed:
Changes to align with ios sdk

avargic committed:
Code review fixes + add validation/extraction methods that take image as byte array

avargic committed:
Code review fixes

avargic committed:
Code review fixes

avargic committed:
Monri SDK version bump
Update readme

avargic committed:
Merge pull request #49 from MonriPayments/feature/scandocTests

avargic committed:
Merge pull request #50 from MonriPayments/feature/scandocTestApp
